### PR TITLE
66 create ndfd ingestion class

### PR DESCRIPTION
**Pull request by bestrada33**
- Run `DBM.insert_lat_lon_test('SBirdIsland', 'South Bird Island', None, '27.4844', '-97.318')` in your `demo.py` first before you run the test.

- Also, ensure that you have activated your virtual environment and installed the new required libraries using:
```python
source ./venv/vsem/Scripts/activate
pip install -r requirements.txt
```

### CODE REVIEW

File: src/DataIngestion/DataIngestion/NDFD.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+            return None
+        except Exception as ex:
+            log(f'Fetch failed, unhandled exceptions: {ex}')
+            return None
+        
+
+    def fetch_predictions(self, seriesRequest: SeriesDescription, timeRequest: TimeDescription, products: str) -> None | dict:
+        url = self.__create_url_pattern(seriesRequest, products)
+
+        response = self.__api_request(url)
```
Comment (matdenkas) [inline src/DataIngestion/DataIngestion/NDFD.py:123]: You might want to add a check here that checks if the response is None, and then returns None

File: src/DataIngestion/DataIngestion/NDFD.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+import os
+SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__)) 
+sys.path.append(os.path.dirname(SCRIPT_DIR))
+
+from DataIngestion.IDataIngestion import IDataIngestion
+from DataClasses import Series, SeriesDescription, Input, TimeDescription
+from SeriesStorage.ISeriesStorage import series_storage_factory
+from utility import log
+
+Time = TypeVar('Time')
```
Comment (matdenkas) [inline src/DataIngestion/DataIngestion/NDFD.py:45]: Can you please go over this tomorrow? This is really cool and I dont think I under stand it

File: src/DataIngestion/DataIngestion/NDFD.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        # specified as 'Z' or '+00:00' ?!
+        #
+        # I'd have to test more to verify, but we just want at least 48 hours of
+        # predictions starting w/ the closest time
+        t0_str = timeRequest.fromDateTime.isoformat()
+        t1_str = timeRequest.toDateTime.isoformat()
+
+        # NOTE:: Needs to be implemented. Also, need to confirm coordinates for SBirdIsland
+        #lat, lon = get_coordinates(seriesRequest.location)
+        if(seriesRequest.dataLocation == 'SBirdIsland'): # NOTE:: Temporary
```
Comment (matdenkas) [inline src/DataIngestion/DataIngestion/NDFD.py:83]: The methodology to get lat and lon is in the db, if you are ready to add it

File: src/DataIngestion/DataIngestion/NDFD.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+            for dataset in self.tree.xpath('/dwml/data/location')[0].getchildren(): latitude = dataset.get('latitude')
+            self.latitude = latitude
+
+        elif time is not None and data is not None:
+            self.time = time
+            self.data = data
+        else:
+            raise ValueError('Either `response` or `time` and `data` must be supplied')
+
+    def zipped(self) -> ZippedDataset[Time, Data]:
```
Comment (matdenkas) [inline src/DataIngestion/DataIngestion/NDFD.py:195]: What black magic is this?????????? This is very cool but looks very complicated

File: src/DataIngestion/DataIngestion/NDFD.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        new_time = dict([(layout, [time_func(t) for t in times]) for (layout, times) in self.time.items()])
+        new_data = [(dataset[0],
+                     dataset[1],
+                     [data_func(value) for value in dataset[2]])
+                    for dataset in self.data]
+        return NDFDPredictions(time=new_time, data=new_data)
+
+
+def iso8601_to_unixms(timestamp: str) -> int:
+    """Convert iso to unix timestamp in milliseconds.  E.g., 2019-04-11T19:00:00-05:00 → 1555027200000"""
```
Comment (matdenkas) [inline src/DataIngestion/DataIngestion/NDFD.py:216]: If you got this code from somewhere you should tag it with @source

File: src/DataIngestion/DataIngestion/NDFD.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+        data: ZippedDataset[int, int] = NDFD_Predictions.map(iso8601_to_unixms, int).zipped()
+
+        # Finally, output data as JSON
+        if len(data) == 1:
+            # If there's only one product, only return the values
+            data_dictionary = json.loads(json.dumps(data[0][1]))
+
+
+        inputs = []
```
Comment (matdenkas) [inline src/DataIngestion/DataIngestion/NDFD.py:135]: You aren't doing anything to filter via an interval. So you would over return over sampled data to SeriesProvider. Im not sure the structure of NDFD if there is no risk of getting oversampled data then this is fine, but if not then we need a clause filtering by an interval here

**Review by matdenkas — CHANGES_REQUESTED** [review]

File: src/DataIngestion/DataIngestion/NDFD.py (hunk 2 above)
Comment (bestrada33) [reply src/DataIngestion/DataIngestion/NDFD.py:45]: I wish I made that. That was Brian or Niall

File: src/DataIngestion/DataIngestion/NDFD.py (hunk 3 above)
Comment (bestrada33) [reply src/DataIngestion/DataIngestion/NDFD.py:83]: Is there an example you can show me?

File: src/DataIngestion/DataIngestion/NDFD.py (hunk 4 above)
Comment (bestrada33) [reply src/DataIngestion/DataIngestion/NDFD.py:195]: Not mine, but I agree

File: src/DataIngestion/DataIngestion/NDFD.py (hunk 5 above)
Comment (bestrada33) [reply src/DataIngestion/DataIngestion/NDFD.py:216]: I put the tag at the very top. Most of this is from Brian. I only added Semaphore related stuff and a couple other additions.

File: src/DataIngestion/DataIngestion/NDFD.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+def iso8601_to_unixms(timestamp: str) -> int:
+    """Convert iso to unix timestamp in milliseconds.  E.g., 2019-04-11T19:00:00-05:00 → 1555027200000"""
+    # lol @ Python's datetime  https://bugs.python.org/issue31800#msg304486
+    #
+    # Anyway, regex kluge since I can't guarantee Python 3.7 and thus
+    # presence of date.fromisoformat() and I'm a hardass who doesn't want
+    # to use dateutil when I think I _shouldn't_ have to
+    time_str = re.sub(r':(\d\d)$', r'\1', timestamp)
+    ms = int(datetime.strptime(time_str, '%Y-%m-%dT%H:%M:%S%z').timestamp()) * 1000
```
Comment (bestrada33) [inline src/DataIngestion/DataIngestion/NDFD.py:226]: Should I remove the multiplication by 1000?

File: src/DataIngestion/DataIngestion/NDFD.py (hunk 6 above)
Comment (bestrada33) [reply src/DataIngestion/DataIngestion/NDFD.py:138]: I don't think interval will ever come in to play for NDFD. The dates are just too all over the place.

File: src/DataIngestion/DataIngestion/NDFD.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+            case 'pAirTemp':
+                return self.fetch_predictions(seriesDescription, timeDescription, 'temp')
+
+
+    def __init__(self):
+        self.sourceCode = "NDFD"
+        self.__seriesStorage = series_storage_factory()
+     
+
+    def __create_url_pattern(self, seriesRequest: SeriesDescription, timeRequest: TimeDescription, product: str) -> str:
```
Comment (lovelysandlonelys) [inline src/DataIngestion/DataIngestion/NDFD.py:70]: do function description? the name is pretty intuitive so maybe it doesn't need one?

**Review by lovelysandlonelys — APPROVED** [review]
most of this is going over my head : D but it seems to work the way it's supposed to

File: src/DataIngestion/DataIngestion/NDFD.py (hunk 8 above)
Comment (bestrada33) [reply src/DataIngestion/DataIngestion/NDFD.py:70]: I probably could. Honestly saw the other ingestion classes didn't have any, so I didn't put any. ¯\_(ツ)_/¯ Adding some now

File: src/DataIngestion/DataIngestion/NDFD.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        if lat is None or lon is None:
+            raise ValueError(f'Empty latlon tuple received in NDFD for locationCode: {seriesRequest.dataLocation}')
+
+        try:
+            beginDate = urllib.parse.quote(t0_str)  # e.g., '2019-04-11T23%3A00%3A00%2B00%3A00'
+            endDate   = urllib.parse.quote(t1_str)
+        except TypeError as e:
+            raise ValueError(f'Error quoting ISO 8601 time string: {e}')
+        
+        base_url = 'https://graphical.weather.gov/xml/sample_products/browser_interface/ndfdXMLclient.php?whichClient=NDFDgen'
```
Comment (abeasley1722) [inline src/DataIngestion/DataIngestion/NDFD.py:103]: I've been told its good practice to add large strings like this to the beginning of the code, but I could be wrong

File: src/tests/test_NDFD.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+load_dotenv()
+
+@pytest.mark.parametrize("source, series, location, interval, datum", [
+    ("NDFD", "pAirTemp", "SBirdIsland", None, None),
+    ("NDFD", "pAirTemp", "SBirdIsland", 3600, None)
+])
+def test_ingest_series(source: str, series: str, location: str, interval: timedelta, datum: str):
+    """This function tests whether each case in the ingest_series
+        function is complete.
+        NOTE:: This test depends on DBM.insert_lat_lon_test('SBirdIsland', 'South Bird Island', None, '27.4844', '-97.318') pre=existing in db
```
Comment (abeasley1722) [inline src/tests/test_NDFD.py:36]: Is it meant to be "pre-existing"?

Comment (bestrada33) [reply src/tests/test_NDFD.py:36]: Yes! "Copying Matthew's comment" moment

File: src/DataIngestion/DataIngestion/NDFD.py (hunk 9 above)
Comment (bestrada33) [reply src/DataIngestion/DataIngestion/NDFD.py:103]: Did they say why?

**Pull request merged**